### PR TITLE
Data transmission protocol definition 

### LINKED ISSUE #2: Data transmission protocol
Write gRPC protocol definition for communication between sensor and server

- Unique bus ID
- Sensed passenger count
- Confidence
- Timestamp

Also assign to @LexiKat (accept the invite!)
cc @kev231998

### PR DESCRIPTION
**Pull request by penguingovernor**
( Closes #2 )

### CODE REVIEW

File: protocols/datagram.proto [hunk 1]
```diff
@@ -0,0 +1,31 @@
+syntax = "proto3";
+
+// TODO:
+// Come up with a better name for this.
+package beavertail;
+
+message DatagramPush {
+  string busID                    = 10;
+  uint64 passengerCount           = 20;
+  double passengerCountConfidence = 30;
```
Comment (natanlao) [inline protocols/datagram.proto]: Do Bryan and KJ know how this is being calculated?

File: protocols/datagram.proto [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  double latitude                 = 40;
+  double longitude                = 50;
+  int64  timestamp                = 60;  // Unix timestamp
```
Comment (natanlao) [inline protocols/datagram.proto]: Should this be
```suggestion
  uint64 timestamp                = 60;  // Unix timestamp
```
?

File: protocols/datagram.proto [hunk 3]
```diff
@@ -0,0 +1,31 @@
+syntax = "proto3";
+
+// TODO:
+// Come up with a better name for this.
+package beavertail;
+
+message DatagramPush {
+  string busID                    = 10;
+  uint64 passengerCount           = 20;
```
Comment (natanlao) [inline protocols/datagram.proto]: Nitpick
```suggestion
  int32 passengerCount            = 20;
```

File: protocols/datagram.proto [hunk 4]
```diff
@@ -0,0 +1,31 @@
+syntax = "proto3";
+
+// TODO:
+// Come up with a better name for this.
+package beavertail;
+
+message DatagramPush {
+  string busID                    = 10;
```
Comment (natanlao) [inline protocols/datagram.proto]: I think the field numbers should all be divided by ten - here's an excerpt from the protobuf documentation

> Note that field numbers in the range 1 through 15 take one byte to encode, including the field number and the field's type (you can find out more about this in Protocol Buffer Encoding). Field numbers in the range 16 through 2047 take two bytes. So you should reserve the numbers 1 through 15 for very frequently occurring message elements. Remember to leave some room for frequently occurring elements that might be added in the future.

**Review by natanlao — CHANGES_REQUESTED** [review]
This looks good! I have some nitpicks about types that I want to address before signing off but overall this is almost ready to merge. I'm a little split, because I'm not convinced that this kind of nitpicking is entirely productive (especially if it only shaves off a couple of bytes per message) but since part of this work will land on sensors, it seems better to take care of this now rather than later.

One other thing that came to mind - that I hadn't yet considered - is that given that the sensors will sometimes go for a long time without being connected to a network, it might be valuable to use the `repeated` clause so that we can cram a bunch of measurements into a single message. But I am not comfortable enough with gRPC to say that assertively, yet. I think we can kick that metaphorical can down the road a little longer.

File: protocols/datagram.proto (hunk 2 above)
Comment (penguingovernor) [reply protocols/datagram.proto]: [Golang](https://golang.org/pkg/time/#Time.Unix) uses int64 for their time stamps so I think int64 should stay, might be different on the client side though

File: protocols/datagram.proto (hunk 1 above)
Comment (penguingovernor) [reply protocols/datagram.proto]: cc: @kev231998 and Brian

File: protocols/datagram.proto (hunk 4 above)
Comment (penguingovernor) [reply protocols/datagram.proto]: Ah didn't see that. Def think this is a good idea.

File: protocols/datagram.proto (hunk 3 above)
Comment (penguingovernor) [reply protocols/datagram.proto]: I like the idea of shrinking the bit count down to 32 bits. However I don't think passengerCount should include negative nums, what do y'all think @kev231998 @LexiKat @kev231998 ?

**Comment by penguingovernor** [comment]
> One other thing that came to mind - that I hadn't yet considered - is that given that the sensors will sometimes go for a long time without being connected to a network, it might be valuable to use the `repeated` clause so that we can cram a bunch of measurements into a single message. But I am not comfortable enough with gRPC to say that assertively, yet. I think we can kick that metaphorical can down the road a little longer.

Perhaps we can have another message type that's like a backlog. I am also scared of the `repeated` keyword but def think this can wait just a tad longer.

File: protocols/datagram.proto (hunk 2 above)
Comment (natanlao) [reply protocols/datagram.proto]: Good point. Both `int64` and `uint64` are variable length anyway, so this change was more for hinting and clarity

File: protocols/datagram.proto (hunk 3 above)
Comment (natanlao) [reply protocols/datagram.proto]: Yikes, I meant uint32

**Review by natanlao — APPROVED** [review]
:+1:

**Comment by natanlao** [comment]
Since @LexiKat is a co-author, I think we can waive reviewer count this one time and merge just so we don't block other work

**Pull request merged**